crates/http/src/macos.rs

```diff
@@ -0,0 +1,179 @@
+use system_configuration::{
+    core_foundation::{
+        array::CFArray,
+        base::CFType,
+        dictionary::CFDictionary,
+        number::CFNumber,
+        string::{CFString, CFStringRef},
+    },
+    dynamic_store::{SCDynamicStore, SCDynamicStoreBuilder, SCDynamicStoreCallBackContext},
+    sys::schema_definitions::{
+        kSCPropNetProxiesHTTPEnable, kSCPropNetProxiesHTTPPort, kSCPropNetProxiesHTTPProxy,
+        kSCPropNetProxiesHTTPSEnable, kSCPropNetProxiesHTTPSPort, kSCPropNetProxiesHTTPSProxy,
+        kSCPropNetProxiesSOCKSEnable, kSCPropNetProxiesSOCKSPort, kSCPropNetProxiesSOCKSProxy,
+    },
+};
+
+pub struct SystemProxiesStore(SCDynamicStore);
+
+#[derive(Default)]
+pub struct SystemProxySettings {
+    pub http: Option<String>,
+    pub https: Option<String>,
+    pub socks: Option<String>,
+}
+
+impl SystemProxySettings {
+    fn from_dynamic_store(store: &SCDynamicStore) -> Self {
+        let Some(proxy_map) = store.get_proxies() else {
+            return Self::default();
+        };
+        SystemProxySettings {
+            http: SystemProxiesStore::get_type(&proxy_map, ProxyType::Http),
+            https: SystemProxiesStore::get_type(&proxy_map, ProxyType::Https),
+            socks: SystemProxiesStore::get_type(&proxy_map, ProxyType::Socks5),
+        }
+    }
+
+    /// prioritize socks over https over http
+    pub fn select(&self) -> Option<String> {
+        if let Some(socks) = &self.socks {
+            Some(socks.as_str())
+        } else if let Some(https) = &self.https {
+            Some(https.as_str())
+        } else if let Some(http) = &self.http {
+            Some(http.as_str())
+        } else {
+            None
+        }
+        .map(str::to_string)
+    }
+}
+
+struct ProxyCallbackInfo<F> {
+    proxy_settings: SystemProxySettings,
+    update_callback: F,
+}
+
+#[derive(Clone, Copy)]
+enum ProxyType {
+    Http,
+    Https,
+    Socks5,
+}
+
+impl SystemProxiesStore {
+    pub fn new<F: FnMut(&SystemProxySettings)>(update_callback: F) -> Self {
```

Comment:
In the future, proxy settings might be lively updated when system settings change OR zed settings change (TODO).